File: python/179. Largest Number.py
#coding=utf-8
__author__ = 'xuxuan'
class Solution(object):
    def largestNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: str
        """
        ans=''
        while len(nums)>0:
            max=self.finMax(nums)
            ans+=str(max)
            nums.remove(max)
        return str(int(ans))
    def compaer(self,n1,n2):
        s1=str(n1)+str(n2)
        s2=str(n2)+str(n1)
        for i in range(len(s1)):
            if s1[i]>s2[i]:
                return True
            if s1[i]<s2[i]:
                return False
        return False
    def finMax(self,nums):
        if len(nums)==1:
            return nums[0]
        max=nums[0]
        for num in nums:
            if self.compaer(num,max):
                max=num
        return max

File: java/179. Largest Number.java
public class Solution {
    public String largestNumber(int[] nums) {
        if (nums.length == 1) {
            return String.valueOf(nums[0]);
        }
        int temp = 0;
        for (int i = nums.length - 1; i > 0; i--) {
            for (int j = 0; j < i; ++j) {
                if (new Solution().compare(nums[j], nums[j + 1])) {
                    temp = nums[j];
                    nums[j] = nums[j + 1];
                    nums[j + 1] = temp;
                }
            }
        }
        StringBuffer sb = new StringBuffer();
        if (nums[nums.length - 1] == 0) {
            return "0";
        }
        for (int i = nums.length - 1; i >= 0; i--) {
            sb.append(String.valueOf(nums[i]));
        }
        return sb.toString();
    }

    public boolean compare(int a, int b) {
        char[] ca=(String.valueOf(a)+String.valueOf(b)).toCharArray();
        char[] cb=(String.valueOf(b)+String.valueOf(a)).toCharArray();
        for (int i = 0; i < ca.length; i++) {
            if (ca[i]>cb[i]) {
                return true;
            }
            if (ca[i]<cb[i]) {
                return false;
            }
        }
        return false;
    }
}